### PR TITLE
Sourcery Starbot ⭐ refactored jobindj/data-science-from-scratch

### PR DESCRIPTION
**Pull request by SourceryAI**
Thanks for starring [sourcery-ai/sourcery](/sourcery-ai/sourcery) ✨ 🌟 ✨

Here's your pull request refactoring your most popular Python repo.

If you want Sourcery to refactor all your Python repos and incoming pull requests [install our bot](/apps/sourcery-ai/installations/new).

<details>
<summary>Review changes via command line</summary>
<p>To manually merge these changes, make sure you're on the <code>master</code> branch, then run:</p>
<pre>
git fetch https://github.com/sourcery-ai-bot/data-science-from-scratch master
git merge --ff-only FETCH_HEAD
git reset HEAD^
</pre>
</details>

### CODE REVIEW

File: first-edition/code/introduction.py [hunk 1]
```diff
@@ -179,9 +179,8 @@ def tenure_bucket(tenure):
 #################
 
 def predict_paid_or_unpaid(years_experience):
-  if years_experience < 3.0: return "paid"
-  elif years_experience < 8.5: return "unpaid"
-  else: return "paid"
+    if years_experience < 3.0 or years_experience >= 8.5: return "paid"
+    else: return "unpaid"
```
Comment (SourceryAI) [inline first-edition/code/introduction.py:183]: Function `predict_paid_or_unpaid` refactored with the following changes:
- Merge duplicate blocks in conditional ([`merge-duplicate-blocks`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#merge-duplicate-blocks))
- Remove redundant conditional ([`remove-redundant-if`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#remove-redundant-if))

File: first-edition/code/natural_language_processing.py [hunk 2]
```diff
@@ -178,7 +178,7 @@ def sample_from(weights):
 
 document_lengths = map(len, documents)
 
-distinct_words = set(word for document in documents for word in document)
+distinct_words = {word for document in documents for word in document}
```
Comment (SourceryAI) [inline first-edition/code/natural_language_processing.py:181]: Lines `181-181` refactored with the following changes:
- Replace list(), dict() or set() with comprehension ([`collection-builtin-to-comprehension`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#collection-builtin-to-comprehension))

File: first-edition/code/statistics.py [hunk 3]
```diff
@@ -35,15 +35,14 @@ def median(v):
...
         return sorted_v[midpoint]
-    else:
-        # if even, return the average of the middle values
-        lo = midpoint - 1
-        hi = midpoint
-        return (sorted_v[lo] + sorted_v[hi]) / 2
+    # if even, return the average of the middle values
+    lo = midpoint - 1
+    hi = midpoint
+    return (sorted_v[lo] + sorted_v[hi]) / 2
```
Comment (SourceryAI) [inline first-edition/code/statistics.py:45]: Function `median` refactored with the following changes:
- Remove unnecessary else after guard condition ([`remove-unnecessary-else`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#remove-unnecessary-else))

File: first-edition/code-python3/clustering.py [hunk 4]
```diff
@@ -126,13 +126,18 @@ def bottom_up_cluster(inputs, distance_agg=min):
...
+                for i, cluster1 in enumerate(clusters)
+                for cluster2 in clusters[:i]
+            ),
+            key=lambda p: cluster_distance(p[0], p[1], distance_agg),
+        )
+
 
         # remove them from the list of clusters
-        clusters = [c for c in clusters if c != c1 and c != c2]
+        clusters = [c for c in clusters if c not in [c1, c2]]
```
Comment (SourceryAI) [inline first-edition/code-python3/clustering.py:140]: Function `bottom_up_cluster` refactored with the following changes:
- Replace multiple comparisons of same variable with `in` operator ([`merge-comparisons`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#merge-comparisons))
- Replace unneeded comprehension with generator ([`comprehension-to-generator`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#comprehension-to-generator))

File: first-edition/code-python3/introduction.py [hunk 5]
```diff
@@ -177,9 +177,8 @@ def tenure_bucket(tenure):
 #################
 
 def predict_paid_or_unpaid(years_experience):
-  if years_experience < 3.0: return "paid"
-  elif years_experience < 8.5: return "unpaid"
-  else: return "paid"
+    if years_experience < 3.0 or years_experience >= 8.5: return "paid"
+    else: return "unpaid"
```
Comment (SourceryAI) [inline first-edition/code-python3/introduction.py:181]: Function `predict_paid_or_unpaid` refactored with the following changes:
- Merge duplicate blocks in conditional ([`merge-duplicate-blocks`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#merge-duplicate-blocks))
- Remove redundant conditional ([`remove-redundant-if`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#remove-redundant-if))

File: first-edition/code-python3/line_count.py [hunk 6]
```diff
@@ -3,9 +3,6 @@
 
 if __name__ == "__main__":
 
-    count = 0
-    for line in sys.stdin:
-        count += 1
-
+    count = sum(1 for _ in sys.stdin)
```
Comment (SourceryAI) [inline first-edition/code-python3/line_count.py:6]: Lines `6-9` refactored with the following changes:
- Replace unused for index with underscore ([`for-index-underscore`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#for-index-underscore))
- Convert for loop into call to sum() ([`sum-comprehension`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#sum-comprehension))

File: first-edition/code-python3/natural_language_processing.py [hunk 7]
```diff
@@ -177,7 +177,7 @@ def sample_from(weights):
 
 document_lengths = [len(d) for d in documents]
 
-distinct_words = set(word for document in documents for word in document)
+distinct_words = {word for document in documents for word in document}
```
Comment (SourceryAI) [inline first-edition/code-python3/natural_language_processing.py:180]: Lines `180-180` refactored with the following changes:
- Replace list(), dict() or set() with comprehension ([`collection-builtin-to-comprehension`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#collection-builtin-to-comprehension))

File: first-edition/code-python3/natural_language_processing.py [hunk 8]
```diff
@@ -217,7 +217,7 @@ def choose_new_topic(d, word):
         topic_word_counts[topic][word] += 1
         topic_counts[topic] += 1
 
-for iter in range(1000):
+for _ in range(1000):
```
Comment (SourceryAI) [inline first-edition/code-python3/natural_language_processing.py:220]: Lines `220-220` refactored with the following changes:
- Replace unused for index with underscore ([`for-index-underscore`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#for-index-underscore))

File: first-edition/code-python3/stats.py [hunk 9]
```diff
@@ -38,11 +38,10 @@ def median(v):
...
         return sorted_v[midpoint]
-    else:
-        # if even, return the average of the middle values
-        lo = midpoint - 1
-        hi = midpoint
-        return (sorted_v[lo] + sorted_v[hi]) / 2
+    # if even, return the average of the middle values
+    lo = midpoint - 1
+    hi = midpoint
+    return (sorted_v[lo] + sorted_v[hi]) / 2
```
Comment (SourceryAI) [inline first-edition/code-python3/stats.py:44]: Function `median` refactored with the following changes:
- Remove unnecessary else after guard condition ([`remove-unnecessary-else`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#remove-unnecessary-else))

File: scratch/crash_course_in_python.py [hunk 10]
```diff
@@ -394,11 +394,7 @@ def some_function_that_returns_a_string():
     return ""
 
 s = some_function_that_returns_a_string()
-if s:
-    first_char = s[0]
-else:
-    first_char = ""
-
+first_char = s[0] if s else ""
```
Comment (SourceryAI) [inline scratch/crash_course_in_python.py:397]: Lines `397-463` refactored with the following changes:
- Simplify numeric comparison ([`simplify-numeric-comparison`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#simplify-numeric-comparison))
- Replace if statement with if expression ([`assign-if-exp`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#assign-if-exp))

File: scratch/deep_learning.py [hunk 11]
```diff
@@ -357,17 +357,17 @@ def backward(self, gradient: Tensor) -> Tensor:
...
+    if not is_1d(tensor):
         return [softmax(tensor_i) for tensor_i in tensor]
 
+    # Subtract largest value for numerical stabilitity.
+    largest = max(tensor)
+    exps = [math.exp(x - largest) for x in tensor]
+
+    sum_of_exps = sum(exps)                 # This is the total "weight".
+    return [exp_i / sum_of_exps             # Probability is the fraction
+            for exp_i in exps]              # of the total weight.
```
Comment (SourceryAI) [inline scratch/deep_learning.py:369]: Function `softmax` refactored with the following changes:
- Swap if/else branches ([`swap-if-else-branches`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#swap-if-else-branches))
- Remove unnecessary else after guard condition ([`remove-unnecessary-else`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#remove-unnecessary-else))

File: scratch/deep_learning.py [hunk 12]
```diff
@@ -402,18 +402,18 @@ def __init__(self, p: float) -> None:
...
             # During evaluation just scale down the outputs uniformly.
             return tensor_apply(lambda x: x * (1 - self.p), input)
 
+        # Create a mask of 0s and 1s shaped like the input
+        # using the specified probability.
+        self.mask = tensor_apply(
+            lambda _: 0 if random.random() < self.p else 1,
+            input)
+        # Multiply by the mask to dropout inputs.
+        return tensor_combine(operator.mul, input, self.mask)
```
Comment (SourceryAI) [inline scratch/deep_learning.py:415]: Function `Dropout.forward` refactored with the following changes:
- Swap if/else branches ([`swap-if-else-branches`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#swap-if-else-branches))
- Remove unnecessary else after guard condition ([`remove-unnecessary-else`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#remove-unnecessary-else))

File: scratch/gradient_descent.py [hunk 13]
```diff
@@ -12,7 +12,7 @@ def difference_quotient(f: Callable[[float], float],
     return (f(x + h) - f(x)) / h
 
 def square(x: float) -> float:
-    return x * x
+    return x**2
```
Comment (SourceryAI) [inline scratch/gradient_descent.py:15]: Function `square` refactored with the following changes:
- Replace x * x with x ** 2 ([`square-identity`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#square-identity))

File: scratch/introduction.py [hunk 14]
```diff
@@ -223,12 +223,10 @@ def tenure_bucket(tenure):
...
-  if years_experience < 3.0:
-    return "paid"
-  elif years_experience < 8.5:
-    return "unpaid"
-  else:
-    return "paid"
+    if years_experience < 3.0 or years_experience >= 8.5:
+        return "paid"
+    else:
+        return "unpaid"
```
Comment (SourceryAI) [inline scratch/introduction.py:229]: Function `predict_paid_or_unpaid` refactored with the following changes:
- Merge duplicate blocks in conditional ([`merge-duplicate-blocks`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#merge-duplicate-blocks))
- Remove redundant conditional ([`remove-redundant-if`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#remove-redundant-if))

File: scratch/network_analysis.py [hunk 15]
```diff
@@ -173,7 +173,7 @@ def page_rank(users: List[User],
     # Small fraction of PageRank that each node gets each iteration
     base_pr = (1 - damping) / num_users
 
-    for iter in tqdm.trange(num_iters):
+    for _ in tqdm.trange(num_iters):
```
Comment (SourceryAI) [inline scratch/network_analysis.py:176]: Function `page_rank` refactored with the following changes:
- Replace unused for index with underscore ([`for-index-underscore`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#for-index-underscore))

File: scratch/neural_networks.py [hunk 16]
```diff
@@ -129,9 +129,9 @@ def fizz_buzz_encode(x: int) -> Vector:
 def binary_encode(x: int) -> Vector:
     binary: List[float] = []
 
-    for i in range(10):
+    for _ in range(10):
         binary.append(x % 2)
-        x = x // 2
+        x //= 2
```
Comment (SourceryAI) [inline scratch/neural_networks.py:134]: Function `binary_encode` refactored with the following changes:
- Replace assignment with augmented assignment ([`aug-assign`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#aug-assign))
- Replace unused for index with underscore ([`for-index-underscore`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#for-index-underscore))

File: scratch/neural_networks.py [hunk 17]
```diff
@@ -153,84 +153,84 @@ def argmax(xs: list) -> int:
 def main():
     import random
     random.seed(0)
-    
+
     # training data
     xs = [[0., 0], [0., 1], [1., 0], [1., 1]]
     ys = [[0.], [1.], [1.], [0.]]
-    
+
```
Comment (SourceryAI) [inline scratch/neural_networks.py:160]: Function `main` refactored with the following changes:
- Replace unused for index with underscore ([`for-index-underscore`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#for-index-underscore))

File: scratch/nlp.py [hunk 18]
```diff
@@ -205,7 +205,7 @@ def sample_from(weights: List[float]) -> int:
 # a list of numbers, one for each document
 document_lengths = [len(document) for document in documents]
 
-distinct_words = set(word for document in documents for word in document)
+distinct_words = {word for document in documents for word in document}
```
Comment (SourceryAI) [inline scratch/nlp.py:208]: Lines `208-208` refactored with the following changes:
- Replace list(), dict() or set() with comprehension ([`collection-builtin-to-comprehension`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#collection-builtin-to-comprehension))

File: scratch/nlp.py [hunk 19]
```diff
@@ -249,7 +249,7 @@ def choose_new_topic(d: int, word: str) -> int:
 
 import tqdm
 
-for iter in tqdm.trange(1000):
+for _ in tqdm.trange(1000):
```
Comment (SourceryAI) [inline scratch/nlp.py:252]: Lines `252-252` refactored with the following changes:
- Replace unused for index with underscore ([`for-index-underscore`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#for-index-underscore))

File: scratch/simple_linear_regression.py [hunk 20]
```diff
@@ -60,36 +60,36 @@ def main():
...
             # Compute loss to stick in the tqdm description
             loss = sum_of_sqerrors(alpha, beta,
                                    num_friends_good, daily_minutes_good)
             t.set_description(f"loss: {loss:.3f}")
-    
+
             # Finally, update the guess
             guess = gradient_step(guess, [grad_a, grad_b], -learning_rate)
-    
+
```
Comment (SourceryAI) [inline scratch/simple_linear_regression.py:92]: Function `main` refactored with the following changes:
- Move assignments closer to their usage ([`move-assign`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#move-assign))

File: scratch/working_with_data.py [hunk 21]
```diff
@@ -581,7 +581,7 @@ def primes_up_to(n: int) -> List[int]:
         with tqdm.trange(3, n) as t:
             for i in t:
                 # i is prime if no smaller prime divides it.
-                i_is_prime = not any(i % p == 0 for p in primes)
+                i_is_prime = all(i % p != 0 for p in primes)
```
Comment (SourceryAI) [inline scratch/working_with_data.py:584]: Function `remove_projection_from_vector.main.primes_up_to` refactored with the following changes:
- Invert any/all to simplify comparisons ([`invert-any-all`](https://github.com/sourcery-ai/sourcery/wiki/Current-Refactorings#invert-any-all))
